### PR TITLE
Lua Binary Trees : Fixed NYI Closure

### PR DESCRIPTION
**Pull request by ExtReMLapin**
Mike pall didn't finish JIT compiling the closures, this is suprising to hear he submited this code.
Also the code he pushed doesn't do like OOP languages, instead of using something like a v-table, it recreate a check function unique for each tree branch.

I moved to what C does.

Perf boost from 8.583000000s to  3.235000000s

We can get twice faster with FFI struct and with disabled garbage collector.

### CODE REVIEW

**Comment by ExtReMLapin** [comment]
If someone's interrested with trying the FFI high speed version (goes from 3.2 from this commit to 2.4 seconds)
here is the code 

```lua

local ffi = require("ffi")

ffi.cdef[[
typedef struct treeBranch treeBranch;
typedef struct treeBranch{
    int item;
    treeBranch *left;
    treeBranch *right;
} treeBranch;
]]

local treeBranch = ffi.typeof("treeBranch")

local branchCounter = 1
local branchList = {}

local function createTreeNode(_item, _left, _right)
  local branch = treeBranch(_item, _left, _right)
  branchList[branchCounter] = branch
  branchCounter = branchCounter + 1
  return branch
end


local checkTreeNode

checkTreeNode = function(branch)
  if branch.left ~= nil then
    return branch.item + checkTreeNode(branch.left) - checkTreeNode(branch.right)
  else
    return branch.item
  end
end
```


I have to store every branch in an array because if you assign a cdata (a ffi struct allocated thru lua with a ffi.typeof constructor) to another struct and not to a lua value, the garbage collector will think it's unused and try to delete it which will lead to segfaults.

You could store it in an hashtable or a list, but slows down the whole thing a lot and makes it useless

**Comment by kostya** [comment]
i prefer not doing such optimizations (and expect interepter to do it under hood). For example, i also can speedup Topaz matmul by 5 times, just by removing closures, it manually unroll code, but this what i expect from interepter (because in real world programs code would be more complex). Nodejs for example have no problem with such similar code, why other interepters, can not optimize it like that?

In other words, interepters should do everything to run average written code fast.

**Comment by kostya** [comment]
btw, thanks i may be add it additionally later as optimized version.

**Comment by ExtReMLapin** [comment]
NodeJS is a web oriented language, web is full of closures because of privacy and async work.

Closures have literally nothing to do in this kind of benchmark, there is a confusion between pseudo-OOP and closures in the code Mike pall produced.

Even if JS really loves and is fine with closures it literally makes no sense to use closures here.

I didn't code JS a lot, but it's clear that even in the JS implementation there is something wrong. [A class needs to be used, not a closure.](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes)

Just because scripting languages is here to help greenhorns/junior developers with programming, doesn't mean we should benchmarks all the languages as if we were all greenhorns.

If you want to can move to from an Argument oriented benchmark to an Object one, while making a Class in Lua, but it won't change change anything in the final results. Either way using closures here or in javascript is a greenhorn mistake.


My point is the following : **It's not an optimization**, it's fixing a developer mistake

------------


In other words, we should either use a class in this implementation to do exactly like v-tables or stick to function and not methods like C does.

Closures have literally nothing to do here


If you want a Class mode (that is slightly slower than what's in the current Pull Request, here it is : )

```lua
-- The Computer Language Benchmarks Game
-- https://salsa.debian.org/benchmarksgame-team/benchmarksgame/
-- contributed by Mike Pall
-- *reset*
local TreeNode = {}
TreeNode.__index = TreeNode

function TreeNode:create(_item, _left, _right)
	local node = {
		item = _item,
		left = _left,
		right = _right
	}

	setmetatable(node, TreeNode)

	return node
end

function TreeNode:check(amount)
	if self.left then
		return self.item + self.left:check() - self.right:check()
	else
		return self.item
	end
end

local function BottomUpTree(item, depth)
	if depth > 0 then
		depth = depth - 1
		local left, right = BottomUpTree(item * 2 - 1, depth), BottomUpTree(item * 2, depth)

		return TreeNode:create(item, left, right)
	else
		return TreeNode:create(item, null, null)
	end
end

start_time = os.clock()
local N = tonumber(arg and arg[1]) or 0
local mindepth = 4
local maxdepth = mindepth + 2

if maxdepth < N then
	maxdepth = N
end

do
	local stretchdepth = maxdepth + 1
	local stretchtree = BottomUpTree(0, stretchdepth)
	io.write(string.format("stretch tree of depth %d\t check: %d\n", stretchdepth, stretchtree:check()))
end

local longlivedtree = BottomUpTree(0, maxdepth)

for depth = mindepth, maxdepth, 2 do
	local iterations = 2 ^ (maxdepth - depth + mindepth)
	local check = 0

	for i = 1, iterations do
		check = check + BottomUpTree(i, depth):check() + BottomUpTree(-i, depth):check()
	end

	io.write(string.format("%d\t trees of depth %d\t check: %d\n", iterations * 2, depth, check))
end

io.write(string.format("long lived tree of depth %d\t check: %d\n", maxdepth, longlivedtree:check()))
io.stderr:write(string.format("time(%.9f)\n", os.clock() - start_time))
```

**Comment by kostya** [comment]
Ok, thanks, I think i see, that this is how oop in lua is. I not really lua familiar.
